### DISCUSSION TITLE
Multiple IR Receiver Support?

### DISCUSSION THREAD

**Discussion by SaltyQuetzals** [Arduino-IRremote/Arduino-IRremote#843] [Q&A] [answered]
Does IRRemote support having multiple receiver input pins? I'd like to identify which IR receiver got a signal, all from the same Arduino board. I saw [this comment](http://www.righto.com/2009/08/multi-protocol-infrared-remote-library.html?showComment=1274069788560#c4007233682250036870) from back in 2010, I was wondering if this has been merged into the library or not. According to the constructors for `IRrecv`, multiple instantiation isn't supported. It would be great to have an example, too, if possible.

**Comment by AnalysIR** [comment, marked as answer]
Can't help with a solution using the library....

**_[deleted old text as previous suggestion  would not work...sorry. However some other hardware solution may work]_**

Can you explain your use case (why you need to know which receiver) as there may be a far simpler method in software, depending on your actual  requirement.

**Reply by SaltyQuetzals** [reply]
I come from a software background, not a hardware/electrical one, so a software solution would be better.

I'm basically trying to put an IR LED on a pair of glasses, and have certain "targets" (IR Receivers) set up around a table. I'd like to know which target the glasses are currently pointed towards at any given time. I am going to try and focus the emission of the IR LED so that it doesn't trigger multiple receivers at once, but that's a problem I can handle after I get an initial system set up.

EDIT: Also, thank you for the quick response!

**Reply by AnalysIR** [reply]
That's much easier, just send different signals thru each LED. (each signals is essentially a binary value) The receiver will then know the source based on on the decoded value.

The bigger problem is that the glasses will receive every signal from every LED unless you have very good physical filtering (like a very directional  pipe) on the glasses end. Also using low power for the emitter will avoid false positives.

Also, some IR receivers will turn off, if too many signals are received in a short space of time. 1 signal every 100->500mSecs might work with the more tolerant receivers.

Be aware that IR is light and will bounce off almost any surface. Assume reflections from literally everywhere.

The issue you will have is more on the glasses side & not focusing on the emitter side. It will really depend on the quantity & angles & distance. look for IR LEDs with narrow beams.

**Reply by AnalysIR** [reply]
...you may get more success sending from the glasses and receiving at the table, but would still need physical filtering.

Have you considered using image recognition using a RPi and OpenCV? Detecting the orientation of a head is pretty straightforward and the camera could be mounted in less obtrusive locations. That is how I would likely approach this problem.

**Reply by SaltyQuetzals** [reply]
I actually _am_ trying to send from glasses and receive at the table, not the other way around! My issue is identifying which receiver got the signal, as the LED won't be controllable once on the glasses. The problem is that IRremote struggles to do multiple receivers at once on the same board, and I'm not sure how I'd do it with multiple boards. The physical filtering is for sure a must, I'm still working out the details on that one.

This setup is actually a prototype of a larger project, and is supposed to require minimal resources. While I think the RPi and OpenCV approach would probably be less-obtrusive, it would require a lot more investment into a prototype that could be discarded the next day.

**Reply by AnalysIR** [reply]
> The problem is that IRremote struggles to do multiple receivers at once on the same board, and I'm not sure how I'd do it with multiple boards. 

As above, that is solved by sending a different signal (code) from each location (a simple binary value).

Another, approach would be to have an emitter and receiver at each location on the table. Then all you would need is a reflective material placed on glasses. _(like the reflective strips on Hi-Vis jackets)_ Then if the location receives back its own signal you get a hit. If 3 out of 10 receive it you may be able to extrapolate the centre point etc. In this scenario physical filtering would also be a must at the table locations & the glasses.

**Reply by SaltyQuetzals** [reply]
> > The problem is that IRremote struggles to do multiple receivers at once on the same board, and I'm not sure how I'd do it with multiple boards.
> 
> As above, that is solved by sending a different signal (code) from each location (a simple binary value).

This is for multiple boards? Or am I missing some component of the IR receiver functionality that would enable this on a single board?

**Reply by AnalysIR** [reply]
If you are receiving at the glasses end, it can be done using a single Arduino board (at the glasses). This would be pretty standard use of IRremote as a receiver.

That is one receiver, one pin and one Arduino. Just use the value decoded for the signal to determine the source.

_...of course you would have to generate a different signal from  each location on the table._

**TIP:** Try using IRremote to record & replay a signal on your TV. You will then understand a bit more of how things work.

**Reply by SaltyQuetzals** [reply]
Okay! Just because of my own individual needs and the amount of work involved in physically filtering four different LEDs instead of just one, I think I'm going to stick with a single IR LED and go with multiple receivers. I will be doing one receiver per board, and then connecting those boards to a primary board which will handle the rest of the logic I want.

**Comment by SaltyQuetzals** [comment]
Solution: Doesn't seem like multiple receivers is something you can do really easily with this library (yet). I am going to have multiple boards, each with their own individual receiver, which will communicate if they received something to a main board which will handle the other logic that I need.
